src/Umbraco.Web/UmbracoContextFactory.cs

```diff
@@ -79,14 +80,26 @@ public UmbracoContextReference EnsureUmbracoContext(HttpContextBase httpContext
                 return new UmbracoContextReference(currentUmbracoContext, false, _umbracoContextAccessor);
 
 
-            httpContext = httpContext ?? new HttpContextWrapper(HttpContext.Current ?? new HttpContext(new SimpleWorkerRequest("null.aspx", "", NullWriterInstance)));
+            httpContext = EnsureHttpContext(httpContext);
 
             var umbracoContext = CreateUmbracoContext(httpContext);
             _umbracoContextAccessor.UmbracoContext = umbracoContext;
 
             return new UmbracoContextReference(umbracoContext, true, _umbracoContextAccessor);
         }
 
+        public static HttpContextBase EnsureHttpContext(HttpContextBase httpContext = null)
+        {
+            if (Thread.GetDomain().GetData(".appPath") is null || Thread.GetDomain().GetData(".appVPath") is null)
```

Comment:
prob best to store the result of `Thread.GetDomain()` instead of re-calling the method. I'll update

Reply:
Not quite sure what this code is doing though, seems to be new?

Reply:
Understandable - If also more or less found the solution using trial-and-error.
After I refactored all of this, we use the fake HttpContext more often in tests.
Some tests fail if we only have the original code
```
return httpContext ?? new HttpContextWrapper(HttpContext.Current ??
                                                         new HttpContext(new SimpleWorkerRequest("null.aspx", "", NullWriterInstance)));
```

because the constructor of `SimpleWorkerRequest` throws a `NullReferenceException`. if `Thread.GetDomain().GetData(".appPath")` or `Thread.GetDomain().GetData(".appVPath")` is `null`.  The constructor is here:
```
    public SimpleWorkerRequest(string page, string query, TextWriter output)
      : this()
    {
      this._queryString = query;
      this._output = output;
      this._page = page;
      this.ExtractPagePathInfo();
      this._appPhysPath = Thread.GetDomain().GetData(".appPath").ToString();
      this._appVirtPath = Thread.GetDomain().GetData(".appVPath").ToString();
      this._installDir = HttpRuntime.AspInstallDirectoryInternal;
      this._hasRuntimeInfo = true;
    }
```

Reply:
Btw - The method is completely removed in https://github.com/umbraco/Umbraco-CMS/pull/7664/, because we there is more abstraction, so we don't need a httpContext directly